src/devtools/client/webconsole/components/Input/JSTerm.tsx

```diff
@@ -0,0 +1,185 @@
+import React, { Dispatch, SetStateAction, useRef, useState } from "react";
+import actions from "devtools/client/webconsole/actions/index";
+import { useDispatch, useSelector } from "react-redux";
+import { useGetRecording, useGetRecordingId } from "ui/hooks/recordings";
+import { getFrameScope } from "devtools/client/debugger/src/reducers/pause";
+import { getCommandHistory } from "../../selectors/messages";
+import {
+  getAutocompleteMatches,
+  getCursorIndex,
+  insertAutocompleteMatch,
+} from "../../utils/autocomplete";
+import Autocomplete from "./Autocomplete";
+import { UIState } from "ui/state";
+import clamp from "lodash/clamp";
+import { WrappedCodeMirror } from "./CodeMirror";
+
+function useGetMatches(expression: string) {
+  const frameScope = useSelector((state: UIState) => getFrameScope(state, "0:0"));
+
+  if (!expression || !frameScope?.scope) {
+    return [];
+  }
+
+  return getAutocompleteMatches(expression, frameScope.scope);
+}
+function useShowAutocomplete(expression: string, hideAutocomplete: boolean) {
+  const matches = useGetMatches(expression);
+  const matchCount = matches.length;
+
+  // Bail if the only suggested autocomplete option has already been applied to the input.
+  if (matchCount === 1 && insertAutocompleteMatch(expression, matches[0]) === expression) {
+    return false;
+  }
+
+  return !hideAutocomplete && matchCount;
+}
+
+function useHistory(setValue: Dispatch<SetStateAction<string>>) {
+  const commandHistory = useSelector(getCommandHistory);
+  const [historyIndex, setHistoryIndex] = useState<number>(0);
+
+  const moveHistoryCursor = (difference: number) => {
```

Comment:
If we always know the difference and want to make sure we don't mess it up, typing it as `difference: -1 | 1` is then even preferred.